### DISCUSSION TITLE
Show a question message box on closing window with option to cancel close

### DISCUSSION THREAD

**Discussion by SimonORorke** [AvaloniaUI/Avalonia#15307] [Q&A] [answered]
In an application for Windows and macOS, I would like to show a question message box on closing the main window, in order to give the user the option to cancel the close.  A handler for the `Window.Closing` event can cancel the close.  So I tried this:
```
private void OnClosing(object? sender, WindowClosingEventArgs e) {
    // Allow the view model to show a message box and return false if the user then opts to cancel the close.
    e.Cancel = !ViewModel.OnClosing();
}
```
That does not work. The problem is that message boxes have to be shown asynchronously. The window closes regardless, and the user never sees the message box.  Making the event handler and the view model method `async` makes no difference, as of course whatever is calling the event handler is not async.

I suspect I'm trying to do the impossible.  However, I would be interested to know a solution, if it exists, or other options.

**Comment by workgroupengineering** [comment, marked as answer]
```c#
private bool _forceClosing = false;

private void OnClosing(object? sender, WindowClosingEventArgs e) {
    // Allow the view model to show a message box and return false if the user then opts to cancel the close.
    if (!_forceClosing)
    {
        e.Cancel = true;
        var dialog = new ConfirmDialog();
        
        dialog.ShowAsync()
        .ContinueWith(t=>
            {
                if (t.Result)
                {
                    _forceClosing =  true;
                    this.Close();
                }
            }, TaskScheduler.FromCurrentSynchronizationContext());
        
    }
}

```

**Comment by SimonORorke** [comment]
That works!  Thank you very much!
